src/mahoji/lib/abstracted_commands/nightmareZoneCommand.ts

```diff
@@ -0,0 +1,482 @@
+		cost: 70,
+		aliases: []
+	},
+	/* TODO: Lock this to max 15 Herb boxes per day
+	{
+		name: 'Herb box',
+		output: new Bank().add('Herb box', 1),
+		cost: 9500,
+		aliases: []
+	},
+	*/
+	{
+		name: 'Vial of water',
+		output: new Bank().add('Vial of water', 1),
+		cost: 145,
+		aliases: []
+	}
+];
+
+export async function nightmareZoneStatsCommand(user: MUser) {
+	const scores = await getMinigameEntity(user.id);
+	return `**Nightmare Zone Stats:**
+
+**Nightmare Zone monsters killed:** ${scores.nmz}.
+**Nightmare Zone points:** ${user.user.nmz_points} Points.`;
+}
+
+export async function nightmareZoneStartCommand(user: MUser, strategy: NMZStrategy, channelID: string) {
+	const skillReqs: Skills = {
+		defence: 70,
+		strength: 70,
+		attack: 70,
+		hitpoints: 70,
+		prayer: 43
+	};
+
+	const [hasReqs] = hasSkillReqs(user, skillReqs);
+	if (!hasReqs) {
+		return `You not meet skill requirements, you need ${Object.entries(skillReqs)
+			.map(([name, lvl]) => `${lvl} ${name}`)
+			.join(', ')}.`;
+	}
+
+	const { QP } = user;
+
+	if (QP < 100) {
+		return `The Nightmare Zone minigame requires **100 QP**, and you have ${QP} QP.\n`;
+	}
+
+	// Check if user have full dharok's
+	if (
+		!user.hasEquippedOrInBank(
+			["Dharok's helm", "Dharok's platebody", "Dharok's platelegs", "Dharok's greataxe"],
+			'every'
+		)
+	) {
+		return "The Nightmare Zone minigame requires full Dharok's equipment for optimal experience/points gained.";
+	}
+
+	const boosts = [];
+
+	let timePerMonster = Time.Minute * 2;
+
+	// combat stat boosts
+	// No monster ID so use -1 as placeholder
+	const [, , attackStyles] = resolveAttackStyles(user, { monsterID: -1 });
+	const skillTotal = sumArr(attackStyles.map(s => user.skillLevel(s))) + user.skillLevel(SkillsEnum.Hitpoints);
+	if (attackStyles.includes(SkillsEnum.Ranged) || attackStyles.includes(SkillsEnum.Magic)) {
+		return 'The Nightmare Zone minigame requires melee combat for efficiency, swap training style using `/minion train style:`';
+	}
+
+	let percent = round(calcWhatPercent(skillTotal, (attackStyles.length + 1) * 99), 2);
+	percent = Math.min(65, (percent - 70) / 0.46);
+	boosts.push(`${percent.toFixed(2)}% for your trained stats (incudling HP)`);
+	timePerMonster = reduceNumByPercent(timePerMonster, percent);
+
+	// Reduce time for item boosts
+	itemBoosts.forEach(set => {
+		for (const item of set) {
+			if (user.hasEquippedOrInBank(item.item.id)) {
+				timePerMonster *= (100 - item.boost) / 100;
+				boosts.push(`${item.boost}% faster for ${item.item.name}`);
+				break;
+			}
+		}
+	});
+
+	const maxTripLength = calcMaxTripLength(user, 'NightmareZone');
+	const quantity = Math.floor(maxTripLength / timePerMonster);
+	const duration = quantity * timePerMonster;
+	// Consume GP (and prayer potion if experience setup)
+	const dreamCost = new Bank().add(
+		'Coins',
+		Math.floor(((QP >= MAX_QP ? 16_000 : 26_000) * duration) / ((strategy === 'points' ? 30 : 60) * Time.Minute))
```

Comment:
This is one very complex line here. I'd probably make a function just for this, and calculate it a bit more imperatively if you can, e.g. same way we do max trip length

```ts
let base = x;
if (something) base /= 2
else base *= 2 / duration
```
otherwise, this line is far too complex to read